File: src/main/java/eladkay/quaritum/common/block/tile/TileEntityBlueprint.java
package eladkay.quaritum.common.block.tile;

import com.google.common.collect.Lists;
import eladkay.quaritum.api.rituals.IDiagram;
import eladkay.quaritum.api.rituals.PositionedBlock;
import eladkay.quaritum.api.rituals.RitualRegistry;
import eladkay.quaritum.common.core.PositionedBlockHelper;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.world.World;

import java.util.List;

public class TileEntityBlueprint extends TileMod {

    public CraftingStage stage = CraftingStage.IDLE;
    public int stageTicks = 0;

    public enum CraftingStage {
        IDLE, PARTICLES, CRAFTING
    }
    public World getWorld() {
        return worldObj;
    }
    @Override
    public void updateEntity() {
        //todo
    }

    private IDiagram getValidRitual() {
        IDiagram bestFit = null;
        int best = -1;
        for (IDiagram ritual : RitualRegistry.getDiagramList()) {
            boolean foundAll = ritual.hasRequiredItems(worldObj, pos, this);
            boolean requirementsMet = ritual.canRitualRun(this.getWorld(), pos, this);
            List<PositionedBlock> blocks = Lists.newArrayList();
            ritual.buildChalks(blocks);
            int chalks = PositionedBlockHelper.getChalkPriority(blocks, this, ritual.getUnlocalizedName());

            if (foundAll && requirementsMet && chalks > best) {
                best = chalks;
                bestFit = ritual;
            }
        }
        runRitual(bestFit);
        return null;
    }

    private boolean runRitual(IDiagram ritual) {
        if(ritual != null && !worldObj.isRemote) {
            ritual.run(worldObj, pos, this);
            return true;
        } else return false;
    }

    public boolean onBlockActivated() {
        return runRitual(getValidRitual());
    }

    @Override
    public void writeCustomNBT(NBTTagCompound compound) {
        compound.setInteger("Stage", stage.ordinal());
        compound.setInteger("StageTicks", stageTicks);
    }

    @Override
    public void readCustomNBT(NBTTagCompound compound) {
        stage = CraftingStage.values()[compound.getInteger("Stage")];
        stageTicks = compound.getInteger("StageTicks");
    }
}

File: src/main/java/eladkay/quaeritum/common/block/tile/TileEntityBlueprint.kt
package eladkay.quaeritum.common.block.tile

import com.teamwizardry.librarianlib.features.autoregister.TileRegister
import com.teamwizardry.librarianlib.features.base.block.tile.TileModTickable
import eladkay.quaeritum.api.rituals.IDiagram
import eladkay.quaeritum.api.rituals.PositionedBlock
import eladkay.quaeritum.api.rituals.RitualRegistry
import eladkay.quaeritum.common.core.PositionedBlockHelper
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing

@TileRegister("blueprint")
class TileEntityBlueprint : TileModTickable() {

    var stage = RitualStage.IDLE
    var stageTicks = 0
    var currentDiagram: IDiagram? = null

    override fun tick() {
        if (!world.isRemote) {
            //LogHelper.logDebug(currentDiagram);
            if (currentDiagram != null)
                if (stageTicks > 0) {
                    stage = RitualStage.PREP
                    stageTicks--
                    if (!currentDiagram!!.onPrepUpdate(world, pos, this, stageTicks))
                        currentDiagram = null
                } else {
                    stage = RitualStage.CRAFTING
                    currentDiagram!!.run(world, pos, this)
                    currentDiagram = null
                }

            if (currentDiagram == null) {
                stage = RitualStage.IDLE
                stageTicks = 0
            }
        }
    }

    private val mirrorStates = arrayOf(-1, 1)

    val bestRitual: IDiagram?
        get() {
            var bestDiagram: IDiagram? = null
            var highestChalks = -1
            diagram@ for (ritual in RitualRegistry.getDiagramList()) {
                for (rotation in EnumFacing.HORIZONTALS) {
                    for (mirror in mirrorStates) {
                        val blocks = arrayListOf<PositionedBlock>()
                        ritual.buildChalks(blocks)
                        PositionedBlockHelper.transform(blocks, rotation, mirror)
                        val chalks = PositionedBlockHelper.getChalkPriority(blocks, this)
                        if (highestChalks < chalks) {
                            bestDiagram = ritual
                            highestChalks = chalks
                            continue@diagram
                        }
                    }
                }
            }
            return bestDiagram
        }

    private fun runRitual(ritual: IDiagram?) {
        if (world.isRemote || ritual == null) return
        currentDiagram = ritual
        stage = RitualStage.PREP
        stageTicks = ritual.getPrepTime(world, pos, this)
    }

    fun onBlockActivated(): Boolean {
        if (currentDiagram == null) {
            val best = bestRitual
            if (best != null) {
                val foundAll = best.hasRequiredItems(world, pos, this)
                val requirementsMet = best.canRitualRun(this.world, pos, this)
                if (foundAll && requirementsMet)
                    runRitual(bestRitual)
            }
        }
        return true
    }

    override fun writeCustomNBT(cmp: NBTTagCompound, sync: Boolean) {
        cmp.setInteger("Stage", stage.ordinal)
        val diagramName = RitualRegistry.getDiagramName(currentDiagram)
        cmp.setString("Diagram", diagramName ?: "")
        cmp.setInteger("StageTicks", stageTicks)
    }

    override fun readCustomNBT(cmp: NBTTagCompound) {
        stage = RitualStage.values()[cmp.getInteger("Stage")]
        val diagramName = cmp.getString("Diagram")
        currentDiagram = if (diagramName == "") null else RitualRegistry.getDiagramByName(diagramName)
        stageTicks = cmp.getInteger("StageTicks")
    }

    enum class RitualStage {
        IDLE, PREP, CRAFTING
    }

}
